### DISCUSSION TITLE
inference with a tensor on yolo11-12

### DISCUSSION THREAD

**Discussion by jimbow973** [ultralytics/ultralytics#19908] [Q&A]
Hi,

Despite all my efforts, it is not impossible to do a prediction with a tensor as input (format BCHW). Any workaround? **The documentation clearly says that the predict method can take a pytorch tensor as input.**
I want later to use a GPU pipeline to feed images to the yolo predictor (DALI, opencv cuda...)

Could you provide a minimum example to predict from a tensor. The sample below returns 0 detections, against 17 detections when using the image path as input instead of the tensor.

```
import torch
from ultralytics import YOLO
import cv2
import numpy as np

# Load the image 
image_path = r"202164537s.png"
cpu_image = cv2.imread(image_path, cv2.IMREAD_COLOR)

# Convert to PyTorch tensor
tensor_image = torch.from_numpy(cpu_image).permute(2, 0, 1).float().div(255.0).unsqueeze(0).to('cuda')

# Load the model
model = YOLO(r"best.engine")

# Run inference
results = model.predict(tensor_image)
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @jimbow973, thank you for bringing this up and for your interest in Ultralytics 🚀!

We recommend checking out the [Docs](https://docs.ultralytics.com/) for detailed guidance on using the `predict` method, including [Python](https://docs.ultralytics.com/usage/python/) examples. It’s great to hear you’re working on a GPU-based pipeline; YOLO is designed to be flexible for such use cases.

If this is a 🐛 Bug Report, could you please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) demonstrating the issue? This will help us better understand and address the problem.

For questions or custom setups ❓, sharing more details like your input tensor structure (e.g., shape, dtype), the YOLO version (YOLO11, YOLO12, etc.), and any error messages or logs will help us assist you more effectively. Also, ensure you've reviewed our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) if applicable.

Join the Ultralytics community to share ideas and get help from fellow users:
- For real-time discussions, visit [Discord](https://discord.com/invite/ultralytics) 🎧.
- For in-depth discussions, check out [Discourse](https://community.ultralytics.com/).
- Engage with the community on our [Subreddit](https://reddit.com/r/Ultralytics).

## Upgrade

Ensure you're using the latest version of `ultralytics` along with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) installed in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). You can upgrade with:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in these verified environments (updated with all dependencies like [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify the correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response, but an Ultralytics engineer will review your question and assist you further shortly. Thank you for your patience! 😊

**Comment by glenn-jocher** [comment]
@jimbow973 for YOLO11 inference with BCHW tensors, ensure your tensor is float32 (0.0-1.0 range) and on the correct device. Here's a minimal example:

```python
from ultralytics import YOLO
import torch

model = YOLO('yolo11n.pt').cuda()  # Load model to GPU
tensor = torch.rand(1, 3, 640, 640, dtype=torch.float32).cuda()  # BCHW tensor
results = model(tensor)  # GPU inference
```

The [Model Prediction with Ultralytics YOLO docs](https://docs.ultralytics.com/modes/predict/) show all supported input formats including PyTorch tensors. For CUDA OpenCV pipelines, ensure proper RGB conversion and normalization before tensor creation.

**Comment by Y-T-G** [comment]
If you're passing a tensor, you need to perform the whole preprocessing manually including resizing.
